File: course/lesson-4/web3py-project/deploy.py
from web3 import Web3
from web3.exceptions import TransactionNotFound
import json
from pathlib import Path

ARTIFACTS_DIR = Path(__file__).parent
ABI_DIR = ARTIFACTS_DIR / "abis"
BYTECODE_DIR = ARTIFACTS_DIR / "artifacts"

def get_abi(contract_name):
    try:
        with open(ABI_DIR / f"{contract_name}.json", 'r') as file:
            return json.load(file)
    except Exception as error:
        print(f"❌ Could not find ABI for contract {contract_name}: {error}")
        raise error

def get_bytecode(contract_name):
    try:
        with open(BYTECODE_DIR / f"{contract_name}.bin", 'r') as file:
            bytecode = file.read().strip()
            return bytecode if bytecode.startswith('0x') else f"0x{bytecode}"
    except Exception as error:
        print(f"❌ Could not find bytecode for contract {contract_name}: {error}")
        raise error

def deploy(config):
    try:
        # Initialize Web3 with RPC URL and longer timeout
        web3 = Web3(Web3.HTTPProvider(
            config["rpc_url"],
            request_kwargs={'timeout': 120}
        ))

        if not config.get("private_key"):
            raise ValueError("PRIVATE_KEY is missing (set it in .env)")

        # Prepare account
        formatted_private_key = config["private_key"] if config["private_key"].startswith('0x') else f"0x{config['private_key']}"
        account = web3.eth.account.from_key(formatted_private_key)
        print(f"Deploying from address: {account.address}")

        chain_id = web3.eth.chain_id
        print(f"Chain ID: {chain_id}")

        # Load ABI and bytecode
        abi = get_abi('Storage')
        bytecode = get_bytecode('Storage')
        print(f"Bytecode length: {len(bytecode)}")

        # Create contract instance
        contract = web3.eth.contract(abi=abi, bytecode=bytecode)

        # Get current nonce (this will test the connection)
        print("Getting nonce...")
        nonce = web3.eth.get_transaction_count(account.address)
        print(f"Nonce: {nonce}")

        balance = web3.eth.get_balance(account.address)
        print(f"Balance: {balance}")

        # Estimate gas using the real constructor call (not raw bytecode only)
        print("Estimating gas...")
        gas_estimate = contract.constructor().estimate_gas({"from": account.address})
        gas_with_buffer = int(gas_estimate * 1.25)
        print(f"Estimated gas: {gas_estimate} (using {gas_with_buffer} with headroom)")

        # Build deployment transaction
        print("Building transaction...")
        construct_txn = contract.constructor().build_transaction({
            'from': account.address,
            'nonce': nonce,
            'chainId': chain_id,
            'gas': gas_with_buffer,
        })

        # Sign transaction
        print("Signing transaction...")
        signed_txn = web3.eth.account.sign_transaction(construct_txn, private_key=formatted_private_key)

        # Send transaction
        print("Sending transaction...")
        tx_hash = web3.eth.send_raw_transaction(signed_txn.raw_transaction)
        print(f"Transaction hash: {tx_hash.hex()}")

        # Wait for transaction receipt
        print("Waiting for transaction receipt...")
        tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash, timeout=300)
        contract_address = tx_receipt.contractAddress

        # Log results
        print(f"✅ Contract deployed at: {contract_address}")
        print(f"Gas used: {tx_receipt.gasUsed}")
        print(f"Block number: {tx_receipt.blockNumber}")

        return web3.eth.contract(address=contract_address, abi=abi)

    except Exception as error:
        print(f'❌ Deployment failed: {error}')
        raise error

if __name__ == "__main__":
    deployment_config = {
        "rpc_url": "http://localhost:8545",
        "private_key": "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5bdc4da702133"
    }

    deploy(deployment_config)

File: homework/homework-4/student-2249/scripts/deploy.ts
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { ethers, JsonRpcProvider } from 'ethers';
import dotenv from "dotenv"
dotenv.config()

// 修复 __dirname
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface ProviderConfig {
  rpc: string;
  chainId: number;
  name: string;
}

const createProvider = (rpcUrl: string, chainId: number, chainName: string) => {
  const provider = new JsonRpcProvider(rpcUrl, {
    chainId: chainId,
    name: chainName,
  });
  return provider;
};

const getAbi = (contractName: string) => {
  try {
    const abiPath = join(__dirname, '../abis', `${contractName}.json`);
    return JSON.parse(readFileSync(abiPath, 'utf8'));
  } catch (error) {
    console.error(`Could not find ABI for contract ${contractName}:`, error instanceof Error ? error.message : String(error));
    throw error;
  }
};

const getByteCode = (contractName: string) => {
  try {
    const bytecodePath = join(__dirname, '../artifacts', `${contractName}.bin`);
    const bytecode = readFileSync(bytecodePath, 'utf8').trim();
    return bytecode.startsWith('0x') ? bytecode : `0x${bytecode}`;
  } catch (error) {
    console.error(`Could not find bytecode for contract ${contractName}:`, error instanceof Error ? error.message : String(error));
    throw error;
  }
};


const deployContract = async (contractName: string, privateKey: string, providerConfig: ProviderConfig) => {
  console.log(`Deploying ${contractName}...`);
  try {
    
    const provider = createProvider(providerConfig.rpc, providerConfig.chainId, providerConfig.name);
    const wallet = new ethers.Wallet(privateKey, provider);
    console.log(`Deploying from address: ${wallet.address}`);

    const factory = new ethers.ContractFactory(getAbi(contractName), getByteCode(contractName), wallet);
    const contract = await factory.deploy();
    await contract.waitForDeployment();

    // Step 3: 保存地址
    const address = await contract.getAddress();
    console.log(`✅ Contract ${contractName} deployed at: ${address}`);

    const addressesFile = join(__dirname, 'contract-address.json');
    const addresses = existsSync(addressesFile) ? JSON.parse(readFileSync(addressesFile, 'utf8')) : {};
    addresses[contractName] = address;
    writeFileSync(addressesFile, JSON.stringify(addresses, null, 2), 'utf8');
    
  } catch (error) {
    console.error(`❌ Failed to deploy contract ${contractName}:`, error);
  }
};

const providerConfig = {
  rpc: 'https://services.polkadothub-rpc.com/testnet',
  chainId: 420420417,
  name: 'polkadot-hub-testnet',
};


const privateKey = dotenv.config().parsed?.PRIVATE_KEY
if (!privateKey) {
  throw new Error("❌ 错误：未找到 PRIVATE_KEY，请检查 .env 文件");
}

deployContract('Storage', privateKey, providerConfig);